### PR TITLE
Usd cameras for production

### PR DESCRIPTION
**Pull request by Vsevolod1983**
We need to be able to select USD camera for production rendering. So we need to load and maintain this list and connect this data to ui elements

### CODE REVIEW

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp [hunk 1]
```diff
@@ -706,19 +723,178 @@ void ProductionSettings::ApplySettings(HdRenderDelegate* renderDelegate)
...
+	MItDependencyNodes iter(MFn::kPluginDependNode);
+	for (; !iter.isDone(); iter.next()) {
+		MObject mobj = iter.item();
+		fn.setObject(mobj);
+		MTypeId id = fn.typeId();
+		MString typeName = fn.typeName();
+		MString str = fn.absoluteName();
+
+		MString origRTypeName = MayaUsdProxyShapeBase::typeName;
+		if (!fn.isFromReferencedFile() && (MayaUsd::LayerManager::supportedNodeType(fn.typeId()))) {
```
Comment (VGorash) [inline lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp:746]: Looks like variables 'id', 'typeName', 'str', 'origTypeName' are unused

**Review by VGorash — APPROVED** [review]

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp [hunk 2]
```diff
@@ -414,22 +417,41 @@ MStatus RprUsdProductionRender::Render()
...
+
+		GfMatrix4d projectionMatrix = camera.GetFrustum().ComputeProjectionMatrix();
+		GfMatrix4d viewMatrix = camera.GetFrustum().ComputeViewMatrix();
 
-	_taskController->SetFreeCameraMatrices(
-		GetGfMatrixFromMaya(viewMatrix.inverse()),
-		GetGfMatrixFromMaya(projMatrix));
+		_taskController->SetFreeCameraMatrices(viewMatrix, projectionMatrix);
+
+		//camera.getmat
```
Comment (DMotorygin) [inline lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp]: Please remove commented out code

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp [hunk 3]
```diff
@@ -486,6 +508,19 @@ void RprUsdProductionRender::RegisterRenderer(const std::string& controlCreation
...
+
+	proc string GetCameraSelectedAttribute()
+	{
+		string $value = `getAttr defaultRenderGlobals.HdRprPlugin_Prod_Static_usdCameraSelected`;
+		return $value;
+	}
+
+	proc SetCameraSelectedAttribute(string $value)
+	{
+		print("Here 333 : " + $value + "\n");
```
Comment (DMotorygin) [inline lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp]: Is this debug message? Is it needed?

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp [hunk 4]
```diff
@@ -520,6 +529,14 @@ std::string ProductionSettings::CreateAttributes()
 
 	std::string controlCreationCmdTemplate = "attrControlGrp -label \"%s\" -attribute \"%s.%s\";";
 
+	// Create attrbiutes for usd camera
+
+	_CreateBoolAttribute(node, MString(g_attributePrefix.GetText()) + "Static_useUSDCamera", false, userDefaults);
+
+	TfTokenVector vec;
+	TfToken token;
+	//_CreateEnumAttribute(node, MString(g_attributePrefix.GetText()) + "Static_usdCameraSelected", vec, token, userDefaults);
```
Comment (DMotorygin) [inline lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp]: Please remove if this line is not needed

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp [hunk 5]
```diff
@@ -706,19 +723,178 @@ void ProductionSettings::ApplySettings(HdRenderDelegate* renderDelegate)
...
+void ProductionSettings::nodeAddedCallback(MObject& node, void* pData)
+{
+	if (_IsOpeningScene)
+	{
+		return;
+	}
+
+	MFnDependencyNode depNode(node);
+
+	MString str1 = depNode.absoluteName();
```
Comment (DMotorygin) [inline lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp]: Are these variables needed?

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp [hunk 6]
```diff
@@ -521,6 +556,107 @@ void RprUsdProductionRender::RegisterRenderer(const std::string& controlCreation
...
+		return $exists; 
+	}
+
+	global proc HdRpr_clearUSDCameras()
+    {
+		global string $g_rprHdrUSDCamerasCtrl;
+		global string $usdCamerasArray[];
+
+		clear($usdCamerasArray);
+		//SetCameraSelectedAttribute("");
```
Comment (DMotorygin) [inline lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp]: Please remove comment if its outdated code

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp [hunk 7]
```diff
@@ -521,6 +556,107 @@ void RprUsdProductionRender::RegisterRenderer(const std::string& controlCreation
...
+
+		$usdCamerasArray[size($usdCamerasArray)] = $cameraName;
+		if (IsUSDCameraCtrlExist())
+		{
+			menuItem -parent $g_rprHdrUSDCamerasCtrl -label $cameraName;
+		}
+
+		if (GetCameraSelectedAttribute() == "")
+		{
+			print("Here 1\n");
```
Comment (DMotorygin) [inline lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp]: Is this debug message? Is it needed?

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp [hunk 8]
```diff
@@ -521,6 +556,107 @@ void RprUsdProductionRender::RegisterRenderer(const std::string& controlCreation
...
+
+		clear($usdCamerasArray);
+		//SetCameraSelectedAttribute("");
+
+		if (!IsUSDCameraCtrlExist())
+			return;
+
+        $items = `optionMenu -q -itemListLong $g_rprHdrUSDCamerasCtrl`;
+        if (size($items) > 0)
+		{
```
Comment (DMotorygin) [inline lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp:628]: Please check formatting in code

Comment (Vsevolod1983) [reply lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp:628]: Done

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp (hunk 7 above)
Comment (Vsevolod1983) [reply lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp]: Done

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp (hunk 6 above)
Comment (Vsevolod1983) [reply lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp]: Done

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp (hunk 5 above)
Comment (Vsevolod1983) [reply lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp]: Done

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp (hunk 4 above)
Comment (Vsevolod1983) [reply lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp]: Done

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp (hunk 3 above)
Comment (Vsevolod1983) [reply lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp]: Done

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp (hunk 2 above)
Comment (Vsevolod1983) [reply lib/mayaUsd/render/mayaToHydra/ProductionRender/RprUsdProductionRender.cpp]: Done

File: lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp (hunk 1 above)
Comment (Vsevolod1983) [reply lib/mayaUsd/render/mayaToHydra/ProductionRender/ProductionSettings.cpp:746]: Done

**Review by DMotorygin — APPROVED** [review]

**Pull request merged**